django/db/backends/base/base.py

```diff
@@ -738,13 +750,42 @@ def on_commit(self, func):
             # immediately.
             func()
 
+    def on_commit_robust(self, func):
+        if not callable(func):
+            raise TypeError("on_commit_robust()'s callback must be a callable.")
+        if self.in_atomic_block:
+            # Transaction in progress; save for execution on commit.
+            self.run_on_commit_robust.append((set(self.savepoint_ids), func))
+        else:
+            # No transaction in progress and in autocommit mode; execute
+            # immediately.
+            try:
+                func()
+            except Exception as e:
+                logger.error(
+                    f"Error calling {func.__qualname__} on_commit_robust()" f" ({e}).",
+                    exc_info=True,
+                )
+
     def run_and_clear_commit_hooks(self):
         self.validate_no_atomic_block()
         current_run_on_commit = self.run_on_commit
         self.run_on_commit = []
+        current_run_on_commit_robust = self.run_on_commit_robust
+        self.run_on_commit_robust = []
         while current_run_on_commit:
             sids, func = current_run_on_commit.pop(0)
             func()
+        while current_run_on_commit_robust:
+            sids, func = current_run_on_commit_robust.pop(0)
+            try:
+                func()
+            except Exception as e:
+                logger.error(
+                    f"Error calling {func.__qualname__} on_commit_robust() "
+                    f"during transaction ({e}).",
+                    exc_info=True,
+                )
```

Comment:
This implies that the `on_commit_robust` hooks are always called **after** the `on_commit` ones.
I'm not sure this is the expected behaviour when we start mixing both 🤔 

Here are some test cases:
```
    def test_multiple_on_commit(self):
        with transaction.atomic():
            self.do(1)
            self.do(2)
            self.do(3)
        self.assertDone([1, 2, 3])

    def test_multiple_on_commit_robust(self):
        with transaction.atomic():
            self.do_robust(1)
            self.do_robust(2)
            self.do_robust(3)
        self.assertDone([1, 2, 3])

    def test_multiple_on_commit_robust_mix(self):
        with transaction.atomic():
            self.do(1)
            self.do_robust(2)
            self.do(3)
        self.assertDone([1, 2, 3])  # fails, because we get 1, 3, 2 currently
```
And I would expect them to be executed in the order they were declared.

One way to implement this, is to register all hooks to the same `self.run_on_commit` list, but instead of having 2-item tuples, append 3-element tuples in the list => `(sid, func, is_robust)` for instance. So you can iterate over them in defined order, and decide whether we try/catch or not depending on `is_robust`.
What do you think?

Reply:
> This implies that the on_commit_robust hooks are always called after the on_commit ones.
> I'm not sure this is the expected behaviour when we start mixing both 🤔

Yes, this is not the expected behaviour,

> One way to implement this, is to register all hooks to the same self.run_on_commit list, but instead of having 2-item tuples, append 3-element tuples in the list => (sid, func, is_robust) for instance. So you can iterate over them in defined order, and decide whether we try/catch or not depending on is_robust.
> What do you think?
Yes, I can implement your suggestion. That will be a nice solution and the functions will be executed in order. I will commit the changes soon. Thank you for the review!